magento/magento2
# Added the possibility of using the Contact module's functionality in headless applications

### Description (*)
As a developer working on backends meant to be used in headless apps like PWA storefronts I would like to be able to utilize the Magento_Contact (the contact form) functionality using GraphQL. 
Currently it's not possible because there's no Magento_ContactGraphQl module that would allow that.

### Related Pull Requests


### Fixed Issues (if relevant)


### Manual testing scenarios (*)
0. As a precondition, configure the Contact feature, go to `/contact/index/` on the regular magento storefront and send the form - you should receive an email. This is not part of this PR, but we'll be doing the same thing via a GraphQL mutation
1. Run the following mutation:
```
mutation {
    submitContactForm(
        input: {
            name: "John Doe",
            email: "johndoe@example.com",
            telephone: "123123123",
            comment: "I like donuts."
        }
    ) {
        success
    }
}
```
2. Test if you received an email and that the response returns `true` for the `success`.

### Questions or comments
As side effects of adding this GraphQL functionality, I did the following things:
1. Improved the re-usability of the email sending feature by moving the validation from the action class to a separate class and the quality of the Post action mostly by adding some return hints and removing the check if it's a POST request, which is obsolete now that the class implements the HttpPostActionInterface
2. The `dev/tests/integration/testsuite/Magento/Contact/Controller/IndexTest.php` test covers all my refactoring in the Magento_Contact module, so I believe that no additional Integration tests are necessary. So I added only API Functional tests to cover the GraphQL part.

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#36830: Added the possibility of using the Contact module's functionality in headless applications
